### PR TITLE
Add validable interface

### LINKED ISSUE #13: Add a `Validable` interface
With a `Validate(ctx) error` method which will be run before an `Update` or a `Save` call.

### PR DESCRIPTION
**Pull request by johnsudaar**
Fix #13

### CODE REVIEW

File: mongo/document/validation_error.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package document
+
+import "bytes"
+
+type ValidationError struct {
+	Errors map[string][]string `json:"errors"`
```
Comment (EtienneM) [inline mongo/document/validation_error.go]: make `Erorrs` private

File: mongo/document/validation_error.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+func NewValidationError() *ValidationError {
```
Comment (EtienneM) [inline mongo/document/validation_error.go]: Don't we want to add an other constructor which takes arugments so that we could write `return NewValidationError("my-field", "errmsg")`?

File: mongo/document/validation_error.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
+package document
+
+import "bytes"
+
+type ValidationError struct {
```
Comment (EtienneM) [inline mongo/document/validation_error.go]: I would prefer calling it `ValidationErrors`

File: mongo/document/validation_error.go (hunk 1 above)
Comment (johnsudaar) [reply mongo/document/validation_error.go]: This will break the json serialization

File: mongo/document/validation_error.go (hunk 2 above)
Comment (johnsudaar) [reply mongo/document/validation_error.go]: We could but i don't see what could be the usage

File: mongo/document/validation_error.go (hunk 3 above)
Comment (johnsudaar) [reply mongo/document/validation_error.go]: :heavy_check_mark:

File: mongo/document/validation_error.go (hunk 2 above)
Comment (EtienneM) [reply mongo/document/validation_error.go]: When there is only one field to validate?

File: mongo/document/base.go [hunk 4]
```diff
@@ -33,10 +33,6 @@ func (d *Base) ensureCreatedAt() {
 	}
 }
 
-func (d *Base) setCreatedAt(t time.Time) {
```
Comment (EtienneM) [inline mongo/document/base.go]: Woot? Why did you remove that?

File: mongo/document/validation_error.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+			buffer.WriteString(err)
+			buffer.WriteString(", ")
+		}
+	}
+	return buffer.String()
+}
+
+// Set will add an error on a specific field, if the field already contains an error, it will just add it to the current errors list
+func (v *ValidationErrors) Set(field, err string) {
+	v.Errors[field] = append(v.Errors[field], err)
```
Comment (EtienneM) [inline mongo/document/validation_error.go]: Don't we need to initialize  `v.Errors[field]` to an empty slice?

File: mongo/document/validation_error.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+// Get will return all errors set for a specific field
+func (v *ValidationErrors) Get(field string) []string {
+	return v.Errors[field]
+}
+
+// Build will send a ValidationErrors struct if there is some errors or nil if no errors has been defined
+func (v *ValidationErrors) Build() *ValidationErrors {
+	if len(v.Errors) == 0 {
```
Comment (EtienneM) [inline mongo/document/validation_error.go]: Shouldn't you be more clever? Like looping through errors to ensure that there is at least an element in an array? I mean, `len(...)` will return 1 for this map:

```go
{
  "my-field": []string{}
}
```
Isn't it?

File: mongo/document/validation_error.go (hunk 5 above)
Comment (EtienneM) [reply mongo/document/validation_error.go]: No we don't

Comment (johnsudaar) [reply mongo/document/validation_error.go]: Nop, we can use the default 0 value

https://play.golang.org/p/dfhSKeElMDE

File: mongo/document/base.go (hunk 4 above)
Comment (johnsudaar) [reply mongo/document/base.go]: Not used

File: mongo/document/validation_error.go (hunk 6 above)
Comment (johnsudaar) [reply mongo/document/validation_error.go]: We could but what's the usecase? How could a user arrive to this position?

Comment (EtienneM) [reply mongo/document/validation_error.go]: As the `Errors` attribute is public, it can happen

Comment (johnsudaar) [reply mongo/document/validation_error.go]: You're righ

File: mongo/document/validation_error.go [hunk 7]
```diff
@@ -41,21 +21,49 @@ func (v *ValidationErrors) Error() string {
...
+//			validations.Set("name", "should not be empty")
+//		}
+//
+//		if m.Email == "" {
+//			validations.Set("email", "should not be empty")
+//		}
+//
+//		return validations.Build()
+//	}
+
```
Comment (EtienneM) [inline mongo/document/validation_error.go]: Remove empty line

File: mongo/document/base.go (hunk 4 above)
Comment (EtienneM) [reply mongo/document/base.go]: `setUpdatedAt` seems also useless, don't you think?

Comment (johnsudaar) [reply mongo/document/base.go]: Nop it's mendatory to implement the `document` interface.
